### ISSUE TITLE
performance issue after upgrade 

### ISSUE THREAD

**Issue by n4mine** [VictoriaMetrics/VictoriaMetrics#1338]
**Describe the bug**
after upgrade vm cluster from 3a8b2eed58df3619409064558f1a1cf8e24e5ecb to a0b001bfeccfd240f7302c61868a49f1a87cd017

1. /api/v1/query_range tp99 rise up
2. io.util reach 100% from 40%
3. cpu usage rise up(mainly `vmstorage`)

**Expected behavior**
performance as usual

**Screenshots**
![image](https://user-images.githubusercontent.com/3345153/119759516-fa837800-beda-11eb-8d91-04bfd63cd8db.png)
![image](https://user-images.githubusercontent.com/3345153/119768142-b9935f80-beea-11eb-8f99-09330c760a6c.png)
![image](https://user-images.githubusercontent.com/3345153/119768166-c57f2180-beea-11eb-932a-c3f81c9bbaa7.png)
![image](https://user-images.githubusercontent.com/3345153/119768195-d29c1080-beea-11eb-9667-5c62ccbd2271.png)
![image](https://user-images.githubusercontent.com/3345153/119768219-db8ce200-beea-11eb-9bcf-5f3d795934b4.png)
![image](https://user-images.githubusercontent.com/3345153/119768270-f1020c00-beea-11eb-86fb-47df42d28cc9.png)
![image](https://user-images.githubusercontent.com/3345153/119768290-fa8b7400-beea-11eb-85c8-2540be2153e4.png)
![image](https://user-images.githubusercontent.com/3345153/119768326-06773600-beeb-11eb-9b7d-afbe72c9a337.png)

**Used command-line flags**
```./vmstorage-prod -storageDataPath /home/xiaoju/data1/vmdata -retentionPeriod 3 -memory.allowedPercent 45 -search.maxUniqueTimeseries 5000000```

```
./vminsert-prod -influxSkipSingleField -insert.maxQueueDuration 10s -replicationFactor=2 -storageNode=......
```

```
./vmselect-prod -search.maxQueryLen 65535 -dedup.minScrapeInterval=1ms -storageNode=......
```

**pprof**

vmstorage:
[cpu.pprof.gz](https://github.com/VictoriaMetrics/VictoriaMetrics/files/6550816/cpu.pprof.gz)
[mem.pprof.gz](https://github.com/VictoriaMetrics/VictoriaMetrics/files/6550817/mem.pprof.gz)

vmselect:
[vmselect.cpu.pprof.gz](https://github.com/VictoriaMetrics/VictoriaMetrics/files/6550973/vmselect.cpu.pprof.gz)
[vmselect.mem.pprof.gz](https://github.com/VictoriaMetrics/VictoriaMetrics/files/6550975/vmselect.mem.pprof.gz)

vminsert:
[vminsert.cpu.pprof.gz](https://github.com/VictoriaMetrics/VictoriaMetrics/files/6550977/vminsert.cpu.pprof.gz)
[vminsert.mem.pprof.gz](https://github.com/VictoriaMetrics/VictoriaMetrics/files/6550978/vminsert.mem.pprof.gz)

**Comment by n4mine** [comment]
**update**

pprof when cpu usage peak

[vmstorage-11-2.cpu.pprof.gz](https://github.com/VictoriaMetrics/VictoriaMetrics/files/6553585/vmstorage-11-2.cpu.pprof.gz)

**Comment by valyala** [comment]
@n4mine , could you build cluster components from the commit 6b90570ed3d43edfa5b902dbb725e416a87abbb1 (or just use the Docker image tag `tags-v1.60.0-cluster-3-g6b90570ed` for cluster components and check whether this helps reducing CPU usage at `vmstorage` nodes?

**Comment by n4mine** [comment]
> build cluster components from the commit 6b90570

it's not help

**Comment by n4mine** [comment]
@valyala 
it's safe to downgrade from 6b90570 to `v1.54.1-cluster`?

is components downgrade sequence matter?

**Comment by n4mine** [comment]
**upgrade**

req duration `vm_request_duration_seconds{job="...", instance="...", quantile="0.99",path="/api/v1/query_range"}`

between enable `-replicationFactor=2`(left) and not(right)
they have same version( a0b001b )

![image](https://user-images.githubusercontent.com/3345153/119935479-245f9c00-bfba-11eb-82dd-7ff0b4d9a982.png)

**Comment by n4mine** [comment]
**update**
after downgrade to `v1.54.1-cluster`, latency became normal

![image](https://user-images.githubusercontent.com/3345153/119957751-f3d92b80-bfd4-11eb-8ec2-9e836857ccf7.png)

**Comment by f41gh7** [comment]
> @valyala
> it's safe to downgrade from [6b90570](https://github.com/VictoriaMetrics/VictoriaMetrics/commit/6b90570ed3d43edfa5b902dbb725e416a87abbb1) to `v1.54.1-cluster`?
> 
> is components downgrade sequence matter?

 Its possible to downgrade. In this case downgrade sequence doesn't matter.


  Few questions: 
1) you upgraded cluster from 1.54.1 to 1.60.0, is it right? 
2) can you provide query examples from api call at vmselect: `/api/v1/status/top_queries` , probably, it will be possible to optimize it.

**Comment by n4mine** [comment]
> you upgraded cluster from 1.54.1 to 1.60.0, is it right?

yes

> can you provide query examples from api call at vmselect: /api/v1/status/top_queries , probably, it will be possible to optimize it.

i has downgrade vm to `v1.54.1`

the follow content base `v1.54.1`, and replace sensitive chars to ***

```
{
  "topN": "20",
  "maxLifetime": "10m0s",
  "search.queryStats.lastQueriesCount": 20000,
  "search.queryStats.minQueryDuration": "0s",
  "topByCount": [
    {
      "accountID": 0,
      "projectID": 0,
      "query": "\\15.count\\-60{__ns__=\"******\",logtype="***",ns="***",url="***",host=\"all\"}",
      "timeRangeSeconds": 3600,
      "count": 5
    },
    {
      "accountID": 0,
      "projectID": 0,
      "query": "\\15.latency\\-60{__ns__=\"******\",host=\"all\",logtype="***",ns="***",url="***"}",
      "timeRangeSeconds": 3600,
      "count": 5
    },
    {
      "accountID": 0,
      "projectID": 0,
      "query": "\\15.success_rate\\-60{__ns__=\"******\",host=\"all\",logtype="***",ns="***",url="***"}",
      "timeRangeSeconds": 3600,
      "count": 4
    },
    {
      "accountID": 0,
      "projectID": 0,
      "query": "\\15.count\\-60{__ns__=\"******\",url="***",host=\"all\",logtype="***",ns="***"}",
      "timeRangeSeconds": 3600,
      "count": 4
    },
    {
      "accountID": 0,
      "projectID": 0,
      "query": "\\1.errcount\\-60{__ns__=\"******\",status=~\"408|499|400\",host=\"all\",logtype="***",ns="***",service="***",url="***"}",
      "timeRangeSeconds": 3600,
      "count": 3
    },
    {
      "accountID": 0,
      "projectID": 0,
      "query": "\\15.latency\\-60{__ns__=\"******\",ns="***",url="***",host=\"all\",logtype="***"}",
      "timeRangeSeconds": 3600,
      "count": 3
    },
    {
      "accountID": 0,
      "projectID": 0,
      "query": "\\15.count\\-60{__ns__=\"******\",host=\"all\",logtype="***",ns="***",url="***"}",
      "timeRangeSeconds": 3600,
      "count": 3
    },
    {
      "accountID": 0,
      "projectID": 0,
      "query": "\\15.count\\-60{__ns__=\"******\",host=\"all\",logtype="***",ns="***",url="***"}",
      "timeRangeSeconds": 3600,
      "count": 3
    },
    {
      "accountID": 0,
      "projectID": 0,
      "query": "\\1.errcount\\-60{__ns__=\"******\",logtype="***",ns="***",service="***",url="***",status=\"596\",host=\"all\"}",
      "timeRangeSeconds": 3600,
      "count": 3
    },
    {
      "accountID": 0,
      "projectID": 0,
      "query": "\\15.count\\-60{__ns__=\"******\",url="***",host=\"all\",logtype="***",ns="***"}",
      "timeRangeSeconds": 3600,
      "count": 3
    },
    {
      "accountID": 0,
      "projectID": 0,
      "query": "\\15.latency\\-60{__ns__=\"******\",ns="***",url="***",host=\"all\",logtype="***"}",
      "timeRangeSeconds": 3600,
      "count": 3
    },
    {
      "accountID": 0,
      "projectID": 0,
      "query": "\\15.errcount\\-60{__ns__=\"******\",logtype="***",ns="***",url="***",status=\"499\",host=\"all\"}",
      "timeRangeSeconds": 3600,
      "count": 3
    },
    {
      "accountID": 0,
      "projectID": 0,
      "query": "\\1.errcount\\-60{__ns__=\"******\",url="***",status=~\"400|408|499\",host=\"all\",logtype="***",ns="***",service="***"}",
      "timeRangeSeconds": 3600,
      "count": 3
    },
    {
      "accountID": 0,
      "projectID": 0,
      "query": "\\15.errcount\\-60{__ns__=\"******\",ns="***",url="***",status=~\"499|504|596\",host=\"all\",logtype="***"}",
      "timeRangeSeconds": 3600,
      "count": 3
    },
    {
      "accountID": 0,
      "projectID": 0,
      "query": "\\15.count\\-60{__ns__=\"******\",host=\"all\",logtype="***",ns="***",url="***"}",
      "timeRangeSeconds": 3600,
      "count": 3
    },
    {
      "accountID": 0,
      "projectID": 0,
      "query": "\\15.errcount\\-60{__ns__=\"******\",host=\"all\",logtype="***",ns="***",url="***",status=~\"499|502|504|557\"}",
      "timeRangeSeconds": 3600,
      "count": 3
    },
    {
      "accountID": 0,
      "projectID": 0,
      "query": "\\1.errcount\\-60{__ns__=\"******\",status=\"596\",host=\"all\",logtype="***",ns="***",service="***",url="***"}",
      "timeRangeSeconds": 3600,
      "count": 2
    },
    {
      "accountID": 0,
      "projectID": 0,
      "query": "\\15.errcount\\-60{__ns__=\"******\",status=\"499\",host=\"all\",logtype="***",ns="***",url="***"}",
      "timeRangeSeconds": 3600,
      "count": 2
    },
    {
      "accountID": 0,
      "projectID": 0,
      "query": "\\15.errcount\\-60{__ns__=\"******\",logtype="***",ns="***",url="***",status=~\"499|504|557\",host=\"all\"}",
      "timeRangeSeconds": 3600,
      "count": 2
    },
    {
      "accountID": 0,
      "projectID": 0,
      "query": "\\1.errcount\\-60{__ns__=\"******\",host=\"all\",logtype="***",ns="***",service="***",url="***",status=~\"502|504|400|408|499|501\"}",
      "timeRangeSeconds": 3600,
      "count": 2
    }
  ],
  "topByAvgDuration": [
    {
      "accountID": 0,
      "projectID": 0,
      "query": "\\1.success_rate\\-60{__ns__=\"******\",ns="***",service="***",url="***",host=\"all\",logtype="***"}",
      "timeRangeSeconds": 25200,
      "avgDurationSeconds": 0.278
    },
    {
      "accountID": 0,
      "projectID": 0,
      "query": "\\10.latency\\-60{__ns__=\"******\",service="***",url="***",host=\"all\",logtype="***"}",
      "timeRangeSeconds": 25200,
      "avgDurationSeconds": 0.202
    },
    {
      "accountID": 0,
      "projectID": 0,
      "query": "\\17.latency\\-60{__ns__=\"******\",logtype="***",url="***",host=\"all\"}",
      "timeRangeSeconds": 25200,
      "avgDurationSeconds": 0.18
    },
    {
      "accountID": 0,
      "projectID": 0,
      "query": "\\60.latency\\-60{__ns__=\"******\",logtype="***",uname="***",service="***",url="***",host=\"all\"}",
      "timeRangeSeconds": 25200,
      "avgDurationSeconds": 0.176
    },
    {
      "accountID": 0,
      "projectID": 0,
      "query": "\\60.count\\-60{__ns__=\"******\",logtype="***",uname="***",service="***",host=\"all\",url="***"}",
      "timeRangeSeconds": 25200,
      "avgDurationSeconds": 0.136
    },
    {
      "accountID": 0,
      "projectID": 0,
      "query": "\\10.latency\\-60{__ns__=\"******\",url="***",host=\"all\",logtype="***",service="***"}",
      "timeRangeSeconds": 25200,
      "avgDurationSeconds": 0.135
    },
    {
      "accountID": 0,
      "projectID": 0,
      "query": "\\60.latency\\-60{__ns__=\"******\",logtype="***",uname="***",service="***",url="***",host=\"all\"}",
      "timeRangeSeconds": 25200,
      "avgDurationSeconds": 0.134
    },
    {
      "accountID": 0,
      "projectID": 0,
      "query": "\\1.count\\-10{__ns__=\"******\",host=\"all\",logtype="***",ns="***",service="***",url="***"}",
      "timeRangeSeconds": 25200,
      "avgDurationSeconds": 0.133
    },
    {
      "accountID": 0,
      "projectID": 0,
      "query": "\\1.latency\\-10{__ns__=\"******\",host=\"all\",ns="***",service="***",url="***",logtype="***"}",
      "timeRangeSeconds": 25200,
      "avgDurationSeconds": 0.126
    },
    {
      "accountID": 0,
      "projectID": 0,
      "query": "\\1.latency\\-10{__ns__=\"******\",host=\"all\",logtype="***",ns="***",service="***",url="***"}",
      "timeRangeSeconds": 25200,
      "avgDurationSeconds": 0.119
    },
    {
      "accountID": 0,
      "projectID": 0,
      "query": "\\60.success_rate\\-60{__ns__=\"******\",service="***",url="***",host=\"all\",logtype="***",uname="***"}",
      "timeRangeSeconds": 25200,
      "avgDurationSeconds": 0.093
    },
    {
      "accountID": 0,
      "projectID": 0,
      "query": "\\10.count\\-60{__ns__=\"******\",host=\"all\",logtype="***",service="***",url="***"}",
      "timeRangeSeconds": 25200,
      "avgDurationSeconds": 0.09
    },
    {
      "accountID": 0,
      "projectID": 0,
      "query": "\\1.latency\\-60{__ns__=\"******\",logtype="***",ns="***",service="***",url="***",host=\"all\"}",
      "timeRangeSeconds": 25200,
      "avgDurationSeconds": 0.089
    },
    {
      "accountID": 0,
      "projectID": 0,
      "query": "\\1.success_rate\\-10{__ns__=\"******\",url="***",host=\"all\",logtype="***",ns="***",service="***"}",
      "timeRangeSeconds": 25200,
      "avgDurationSeconds": 0.089
    },
    {
      "accountID": 0,
      "projectID": 0,
      "query": "\\52.latency\\-60{__ns__=\"******\",uname="***",logtype="***",url="***",host=\"all\"}",
      "timeRangeSeconds": 25200,
      "avgDurationSeconds": 0.087
    },
    {
      "accountID": 0,
      "projectID": 0,
      "query": "\\60.latency\\-60{__ns__=\"******\",host=\"all\",logtype="***",uname="***",service="***",url="***"}",
      "timeRangeSeconds": 25200,
      "avgDurationSeconds": 0.086
    },
    {
      "accountID": 0,
      "projectID": 0,
      "query": "\\1.count\\-60{__ns__=\"******\",host=\"all\",ns="***",service="***",url="***",logtype="***"}",
      "timeRangeSeconds": 25200,
      "avgDurationSeconds": 0.086
    },
    {
      "accountID": 0,
      "projectID": 0,
      "query": "\\60.latency\\-60{__ns__=\"******\",service="***",url="***",host=\"all\",logtype="***",uname="***"}",
      "timeRangeSeconds": 25200,
      "avgDurationSeconds": 0.083
    },
    {
      "accountID": 0,
      "projectID": 0,
      "query": "\\52.success_rate\\-60{__ns__=\"******\",url="***",host=\"all\",uname="***",logtype="***"}",
      "timeRangeSeconds": 25200,
      "avgDurationSeconds": 0.075
    },
    {
      "accountID": 0,
      "projectID": 0,
      "query": "\\1.success_rate\\-10{__ns__=\"******\",url="***",logtype="***",host=\"all\",ns="***",service="***"}",
      "timeRangeSeconds": 25200,
      "avgDurationSeconds": 0.069
    }
  ],
  "topBySumDuration": [
    {
      "accountID": 0,
      "projectID": 0,
      "query": "\\1.success_rate\\-60{__ns__=\"******\",ns="***",service="***",url="***",host=\"all\",logtype="***"}",
      "timeRangeSeconds": 25200,
      "sumDurationSeconds": 0.278
    },
    {
      "accountID": 0,
      "projectID": 0,
      "query": "\\10.latency\\-60{__ns__=\"******\",service="***",url="***",host=\"all\",logtype="***"}",
      "timeRangeSeconds": 25200,
      "sumDurationSeconds": 0.202
    },
    {
      "accountID": 0,
      "projectID": 0,
      "query": "\\17.latency\\-60{__ns__=\"******\",logtype="***",url="***",host=\"all\"}",
      "timeRangeSeconds": 25200,
      "sumDurationSeconds": 0.18
    },
    {
      "accountID": 0,
      "projectID": 0,
      "query": "\\60.latency\\-60{__ns__=\"******\",logtype="***",uname="***",service="***",url="***",host=\"all\"}",
      "timeRangeSeconds": 25200,
      "sumDurationSeconds": 0.176
    },
    {
      "accountID": 0,
      "projectID": 0,
      "query": "\\60.count\\-60{__ns__=\"******\",logtype="***",uname="***",service="***",host=\"all\",url="***"}",
      "timeRangeSeconds": 25200,
      "sumDurationSeconds": 0.136
    },
    {
      "accountID": 0,
      "projectID": 0,
      "query": "\\10.latency\\-60{__ns__=\"******\",url="***",host=\"all\",logtype="***",service="***"}",
      "timeRangeSeconds": 25200,
      "sumDurationSeconds": 0.135
    },
    {
      "accountID": 0,
      "projectID": 0,
      "query": "\\60.latency\\-60{__ns__=\"******\",logtype="***",uname="***",service="***",url="***",host=\"all\"}",
      "timeRangeSeconds": 25200,
      "sumDurationSeconds": 0.134
    },
    {
      "accountID": 0,
      "projectID": 0,
      "query": "\\1.count\\-10{__ns__=\"******\",host=\"all\",logtype="***",ns="***",service="***",url="***"}",
      "timeRangeSeconds": 25200,
      "sumDurationSeconds": 0.133
    },
    {
      "accountID": 0,
      "projectID": 0,
      "query": "\\1.latency\\-10{__ns__=\"******\",host=\"all\",ns="***",service="***",url="***",logtype="***"}",
      "timeRangeSeconds": 25200,
      "sumDurationSeconds": 0.126
    },
    {
      "accountID": 0,
      "projectID": 0,
      "query": "\\1.latency\\-10{__ns__=\"******\",host=\"all\",logtype="***",ns="***",service="***",url="***"}",
      "timeRangeSeconds": 25200,
      "sumDurationSeconds": 0.119
    },
    {
      "accountID": 0,
      "projectID": 0,
      "query": "\\60.success_rate\\-60{__ns__=\"******\",service="***",url="***",host=\"all\",logtype="***",uname="***"}",
      "timeRangeSeconds": 25200,
      "sumDurationSeconds": 0.093
    },
    {
      "accountID": 0,
      "projectID": 0,
      "query": "\\10.count\\-60{__ns__=\"******\",host=\"all\",logtype="***",service="***",url="***"}",
      "timeRangeSeconds": 25200,
      "sumDurationSeconds": 0.09
    },
    {
      "accountID": 0,
      "projectID": 0,
      "query": "\\1.latency\\-60{__ns__=\"******\",logtype="***",ns="***",service="***",url="***",host=\"all\"}",
      "timeRangeSeconds": 25200,
      "sumDurationSeconds": 0.089
    },
    {
      "accountID": 0,
      "projectID": 0,
      "query": "\\1.success_rate\\-10{__ns__=\"******\",url="***",host=\"all\",logtype="***",ns="***",service="***"}",
      "timeRangeSeconds": 25200,
      "sumDurationSeconds": 0.089
    },
    {
      "accountID": 0,
      "projectID": 0,
      "query": "\\52.latency\\-60{__ns__=\"******\",uname="***",logtype="***",url="***",host=\"all\"}",
      "timeRangeSeconds": 25200,
      "sumDurationSeconds": 0.087
    },
    {
      "accountID": 0,
      "projectID": 0,
      "query": "\\60.latency\\-60{__ns__=\"******\",host=\"all\",logtype="***",uname="***",service="***",url="***"}",
      "timeRangeSeconds": 25200,
      "sumDurationSeconds": 0.086
    },
    {
      "accountID": 0,
      "projectID": 0,
      "query": "\\1.count\\-60{__ns__=\"******\",host=\"all\",ns="***",service="***",url="***",logtype="***"}",
      "timeRangeSeconds": 25200,
      "sumDurationSeconds": 0.086
    },
    {
      "accountID": 0,
      "projectID": 0,
      "query": "\\60.latency\\-60{__ns__=\"******\",service="***",url="***",host=\"all\",logtype="***",uname="***"}",
      "timeRangeSeconds": 25200,
      "sumDurationSeconds": 0.083
    },
    {
      "accountID": 0,
      "projectID": 0,
      "query": "\\52.success_rate\\-60{__ns__=\"******\",url="***",host=\"all\",uname="***",logtype="***"}",
      "timeRangeSeconds": 25200,
      "sumDurationSeconds": 0.075
    },
    {
      "accountID": 0,
      "projectID": 0,
      "query": "\\1.success_rate\\-10{__ns__=\"******\",url="***",logtype="***",host=\"all\",ns="***",service="***"}",
      "timeRangeSeconds": 25200,
      "sumDurationSeconds": 0.069
    }
  ]
}
```

**Comment by n4mine** [comment]
**update**

![image](https://user-images.githubusercontent.com/3345153/119976222-5fc58f00-bfe9-11eb-97ac-2dbbb8cbc821.png)

**Comment by f41gh7** [comment]
Thanks a lot, we will take a look.

**Comment by valyala** [comment]
@n4mine , could you try `vmstorage` from the commit 1e4a64844d66413205bb0bf7ee5785b021750365 ?

**Comment by n4mine** [comment]
> try vmstorage from the commit 1e4a648

it doesn't work

query duration slower and io.util rise to 100%

**Comment by n4mine** [comment]
after upgrade, `threads` and `active merges`  going up, it's related?

![image](https://user-images.githubusercontent.com/3345153/121175889-2ebd4800-c88e-11eb-9423-4702819614af.png)
![image](https://user-images.githubusercontent.com/3345153/121176210-8956a400-c88e-11eb-997c-4f66c0fe454b.png)

**Comment by n4mine** [comment]
**update** 

new clues

the rate of `vm_filestream_real_write_calls_total` which upgrade from `v1.54.1-cluster` was rise up

![image](https://user-images.githubusercontent.com/3345153/121275252-d1adaa80-c8fe-11eb-917e-13312cb9a831.png)

**Comment by n4mine** [comment]
@valyala 

i think it's related to fc902734d9ac36e6148b752bc38746246d5f6b69
after upgrade vmstorage from 30ab0eec27648c8ecac45896fff7a5d65ffef58a to fc902734d9ac36e6148b752bc38746246d5f6b69

the query duration rise up

**Comment by valyala** [comment]
@n4mine , could you try tuning the `loopsCountPerMetricNameMatch` constant at the following line: https://github.com/VictoriaMetrics/VictoriaMetrics/blob/5f91a701fa11d1a089331d2babe60a1d9406b429/lib/storage/index_db.go#L3288 

Try increasing it to 200, 300, 400, ..., etc. until the CPU usage for your workload becomes similar to the commit https://github.com/VictoriaMetrics/VictoriaMetrics/commit/30ab0eec27648c8ecac45896fff7a5d65ffef58a .

**Comment by n4mine** [comment]
@valyala 

https://snapshot.raintank.io/dashboard/snapshot/lB6lLgMXLSbfs2306sY96d5T34W1LCf3

the `vmstorage-11-sf-641df-1.docker.ys` is the node which set `loopsCountPerMetricNameMatch` to 200, 300, 400, 500, 600 and 700
and the `vmstorage-11-sf-641df-19.docker.ys` is the node which deployed the version 3a8b2eed

* 20:28 +8000: 200
* 20:53 +8000: 300
* 21:17 +8000: 400
* 21:33 +8000: 500
* 22:06 +8000: 600
* 22:21 +8000: 700
* 22:34 +8000: 30ab0eec

i think it's not obvious to adjust `loopsCountPerMetricNameMatch`
at 22:34, notice the `query duration`, `cpu usage` and `cpu idle` after 30ab0eec deployed, everything is back

**Comment by wf1nder** [comment]
We are facing with similar issue.
Upgrade vmstorage 1.58.0 > 1.59.0 causes infinite indexdb merges, excessive IO load, load average over 115. CPU consumption looks almost as on 1.58.0 at the same time. On downgrade to 1.58.0 everythings returns to normal.

If we turn off writing to cluster (stop vminsert) extensive merges and IO load is declining almost to 0, so this is not some background operations.

Looks like this is related to https://github.com/VictoriaMetrics/VictoriaMetrics/issues/1244

The problem is reproducing stably on whole 3 vmstorage shards in cluster. Each server have 104 virtual cpu cores.
I tried to set GOMAXPROCS variable to value 8, and this to some extent solves the problem. But load declining not to values at 1.58.0 version, indexdb merges are still performed more often. And generally I'm not sure it's a good solution.

So we are stuck now on the version 1.58.0

**Comment by valyala** [comment]
@n4mine , @wf1nder ,  thanks for the assistance and for the additional details!

The commit aa9b56a046b6ae8083fa659df35dd5e994bf9115 should help reducing disk write usage to the levels comparable to the commit prior to https://github.com/VictoriaMetrics/VictoriaMetrics/commit/2d1d60118d0d49e2bc98f19ed4b1acecd2048975 . It would be great if you could verify this commit on your workloads and post results here.

As for the commit https://github.com/VictoriaMetrics/VictoriaMetrics/commit/fc902734d9ac36e6148b752bc38746246d5f6b69 , which resulted in higher CPU usage on queries from @n4mine 's workload, it would be great if higher values for `loopsCountPerMetricNameMatch` could be tested as well until the CPU usage returns the levels prior the commit fc902734d9ac36e6148b752bc38746246d5f6b69 . We cannot simply revert to the commit prior to fc902734d9ac36e6148b752bc38746246d5f6b69, since the commit fc902734d9ac36e6148b752bc38746246d5f6b69 belongs to a series of commits, which resulted in reduced CPU usage and to smaller query latency on a wide range of production queries. The @n4mine 's workload workload hits a regression, which must be identified and fixed.

By default VictoriaMetrics locates internal series ids (aka `metricID` in the source code) for every label filter used in the query. E.g. a query `foo{bar="baz"}` contains two label filters: `{__name__="foo"}` (which can be simply written as `foo`) and `{bar="baz"}`. So VictoriaMetrics finds a set for series ids with the name `foo` and a set for series ids with the label `{bar="baz"}`. Then it finds the intersection of these sets. But sometimes it is faster to switch to another algorithm if a particular label filter matches too many series. For example, if every series contains a label `{env="prod"}` and the query contains the identical filter - `{env="prod"}`, then such filter would select all series. The number of selected series can be big (e.g. hundreds of millions) on workloads with high number of active time series and a high churn rate. In this case it is better switching to the following algorithm if the number of series found via other label filters is quite low:
1. To select all the labels for the found series
2. To apply the remaining label filters directly to the labels of the found series on the step 1.

Unfortunately, the @n4mine 's workload triggers the alternative algorithm too early according to the provided cpu profiles, so the alternative algorithm takes much more CPU time than the original algorithm. The `loopsCountPerMetricNameMatch` constant is used as a threshold for switching between these algorithms. So it would be great to find the good value for the  `loopsCountPerMetricNameMatch` constant for @n4mine 's workload. It definitely must be bigger than 700 according to https://github.com/VictoriaMetrics/VictoriaMetrics/issues/1338#issuecomment-860756224 .

**Comment by valyala** [comment]
oops, there was a bug in the commit aa9b56a046b6ae8083fa659df35dd5e994bf9115 , which could result to an infinite loop on high ingestion rate. The bug has been fixed in the commit dcbc22552f0dcfa3075982eabeda44b0a28fd428 (single-node) and a207be3ffb6f196fd3826b65ab273604cdd99431 (cluster version).

**Comment by wf1nder** [comment]
@valyala Just builded and tested vmstorage from master, which includes changes from https://github.com/VictoriaMetrics/VictoriaMetrics/commit/aa9b56a046b6ae8083fa659df35dd5e994bf9115, and now everything is fine! IO consumption, load average and indexdb merge frequency are the same as on 1.58.0.
Thanks a lot!

**Comment by n4mine** [comment]
@valyala 

the io problems has been solved. 

> higher values for `loopsCountPerMetricNameMatch`

i would like to test this.
and what should i increase the `loopsCountPerMetricNameMatch`, step by 100 each time?

**Comment by valyala** [comment]
> what should i increase the loopsCountPerMetricNameMatch, step by 100 each time?

Try using binary search for finding the minimum value, which results in reduced CPU usage for the given workload. For example, start with quite big value such as 10000 or even higher, make sure the CPU usage has been reduced and then decrease the value until the CPU usage starts increasing again.

**Comment by n4mine** [comment]
@valyala 

i think it's not obvious for reduce cpu usage by adjust `loopsCountPerMetricNameMatch `

23:11(utc) 100
23:29 200
23:51 500
00:06 1000
00:24 2000
00:33 3000
00:51 4000
00:59 5000
02:23 6000
02:50 7000
03:42 8000
04:32 9000
04:45 10000
05:35 20000
05:57 50000
08:33 100000

https://snapshot.raintank.io/dashboard/snapshot/T7KnL44U8znXyMcJarAm93mEMVzKIhGo

**Comment by valyala** [comment]
> and even i set loopsCountPerMetricNameMatch to 1000000, it's not working

This is unexpected :( Could you share a cpu profile collected from VictoriaMetrics with `loopsCountPerMetricNameMatch=100000`? It would be great if the system could work for more than an hour with this value for `loopsCountPerMetricNameMatch` before collecting the cpu profile. This should eliminate possible observation distortion related to the cache.

**Comment by n4mine** [comment]
> cpu profile

[11-19.pprof.gz](https://github.com/VictoriaMetrics/VictoriaMetrics/files/6676957/11-19.pprof.gz)

**Comment by valyala** [comment]
> i think it's not obvious for reduce query duration and cpu usage by adjust loopsCountPerMetricNameMatch

Agreed. The `loopsCountPerMetricNameMatch` is an implementation detail, which shouldn't be visible to user. Unfortunately its value doesn't work as expected for your workload. That's why we need determining the value suitable for your workload, so the constant and / or the algorithm could be adjusted in the next release of VictoriaMetrics.

Thanks for the provided cpu profile! It shows that the code didn't switch to metricName match as expected for high value of `loopsCountPerMetricNameMatch`.

@n4mine , could you build and run `vmstorage` from the commit 0198a14b581f0a481c3ee21b34a9b869a6b77c6d and then post here a sample of log lines with the `debug: switching to metricName match for the remaining` prefix?

**Comment by n4mine** [comment]
> post here a sample of log lines with the debug: switching to metricName match for the remaining prefix?

[vmstorage.log.zip](https://github.com/VictoriaMetrics/VictoriaMetrics/files/6679482/vmstorage.log.zip)

**Comment by valyala** [comment]
@n4mine , thanks for the provided logs. Could you build `vmstorage` from the commit a22f37599b7fd469dc4873b7f1fad670dd0e16a4 and verify whether it helps reducing high CPU usage on your workload?

**Comment by n4mine** [comment]
Finally, everything has recovery. thanks @valyala

**Comment by n4mine** [comment]
Finally, everything is back to normal. 
thanks @valyala

**Comment by valyala** [comment]
The commit which reduces disk write IO usage and the commit which reduces CPU usage are included in [v1.62.0](https://github.com/VictoriaMetrics/VictoriaMetrics/releases). Closing this issue as fixed.
